### ISSUE TITLE
OIDC authentication via Google fails on Tomcat

### ISSUE THREAD

**Issue by tfield** [micronaut-projects/micronaut-servlet#728]
### Expected Behavior

When using OIDC authentication with Google and Tomcat, I expect to be able to log in successfully.

The Micronaut Guide SECURE A MICRONAUT APPLICATION WITH GOOGLE was followed, and the server was changed from netty to tomcat, and this fails.

This seems similar to
https://github.com/micronaut-projects/micronaut-servlet/issues/239
which includes links to
https://github.com/micronaut-projects/micronaut-servlet/pull/256
https://github.com/micronaut-projects/micronaut-servlet/pull/275

### Actual Behaviour

The callback URL fails with the following exception 

`ERROR i.m.http.server.RouteExecutor - Unexpected error occurred: Error instantiating bean of type [io.micronaut.security.oauth2.endpoint.authorization.response.DefaultOpenIdAuthorizationResponse]: Error decoding request body: Stream closed`

This works fine with Netty.

[mn-tomcat-google-error.txt](https://github.com/user-attachments/files/15522761/mn-tomcat-google-error.txt)

### Steps To Reproduce

1. download the java/gradle micronaut guide sample from [here](https://guides.micronaut.io/latest/micronaut-oauth2-oidc-google.html).
2. configure google client id/secret per your google console
3. start and test the app to authenticate
4. observe success
5. edit build.gradle and change micronaut.runtime from 'netty' to 'tomcat', and add `implementation("io.micronaut.servlet:micronaut-http-server-tomcat")`
6. restart the app and test to authenticate
7. observe the stack trace

### Environment Information

- macOS Sonoma
- azul 21 jdk

### Example Application

Download the java/gradle micronaut guide sample from [here](https://guides.micronaut.io/latest/micronaut-oauth2-oidc-google.html).

### Version

4.4.0

**Comment by graemerocher** [comment]
Seems it doesn't like async being enabled by default, will take a look. You could try and use the latest 4.9.0 version and disable async. See https://micronaut-projects.github.io/micronaut-servlet/latest/guide/configurationreference.html#io.micronaut.servlet.engine.MicronautServletConfiguration

**Comment by tfield** [comment]
Thank you - that worked!  I set

```
    implementation 'io.micronaut.servlet:micronaut-servlet-engine:4.9.0'
    implementation 'io.micronaut.servlet:micronaut-http-server-tomcat:4.9.0'
```

and

```
micronaut:
  servlet:
    async-supported: false
    test-async-supported: false
    async-file-serving-enabled: false
```

and this was successful.

**Comment by tfield** [comment]
Update: I have updated my app to 4.6.3 and doing so allows me to remove the servlet 4.9 dependencies, so that's great.
However, when I bump to 4.7.0, OIDC authentication no longer works. Is it possible to track down what has changed that broke this?

**Comment by graemerocher** [comment]
Validation notes from local repro on 2026-04-29.

I used the current Google OIDC guide sample, switched it from Netty to Tomcat (same change described in the issue), and then exercised the callback route directly with a synthetic form-post request to isolate the reported failure point.

Test shape:
- downloaded `micronaut-oauth2-oidc-google-gradle-java.zip` from the guide site
- changed the sample to `runtime("tomcat")` and added `implementation("io.micronaut.servlet:micronaut-http-server-tomcat")`
- started the app with dummy OAuth settings (`OAUTH_CLIENT_ID`, `OAUTH_CLIENT_SECRET`, `OAUTH_ISSUER=http://127.0.0.1:9999`)
- sent:

```bash
curl -i -X POST http://127.0.0.1:<port>/oauth/callback/google \
  -H 'Content-Type: application/x-www-form-urlencoded' \
  --data 'code=fake-code&state=fake-state&scope=openid%20email%20profile'
```

Versions checked:
- Micronaut 4.4.0 / Tomcat 10.1.20
- Micronaut 4.6.3 / Tomcat 10.1.28
- Micronaut 4.7.0 / Tomcat 10.1.31
- current guide sample line (Micronaut 4.10.9) / Tomcat 11.0.11

Observed result in all cases:
- the callback request reached Micronaut Security state handling and failed later with `Illegal base64 character 5b` because the synthetic `state` value was intentionally fake
- I did **not** reproduce the original symptom `Error decoding request body: Stream closed`
- in other words, request processing advanced past body decoding / callback-body binding instead of failing at `DefaultOpenIdAuthorizationResponse`

Because the reported failure point did not reproduce in this focused callback test matrix, I don’t currently have evidence that this is still actionable in supported versions. The remaining gap is that I did not run a live Google round-trip with real client credentials, but the direct callback probe exercises the route where the reported `Stream closed` exception was occurring.

Based on this validation, I’m marking this as `closed: cannot reproduce` and leaving it for human review.
